### ISSUE TITLE
KL Divergence

### ISSUE THREAD

**Issue by zhangchuanyi96** [dougbrion/pytorch-classification-uncertainty#3]
Hi, I am very interested in this work. 

I find that `kl_divergence` loss in the code seems differ from that in the paper. 
`
    lnB_uni = torch.sum(torch.lgamma(beta), dim=1,
                        keepdim=True) - torch.lgamma(S_beta)
`

`torch.sum(torch.lgamma(beta), dim=1,  keepdim=True)` for this part, I can't find it in the paper. Maybe I misunderstand it. Can you help me?

**Comment by iamwinter** [comment]
I was gonna ask the same question!
In the another [repo](https://github.com/atilberk/evidential-deep-learning-to-quantify-classification-uncertainty/blob/master/CMPE547_TermProject.ipynb) also appear this problem. 

By the way, it's known that gamma(K)=1 where K is a vertor of K ones. Then, why not ignore this element which located at denominator?  Therefore,  I think it likes both repositories mentioned above show the wrong implementations.

Did you help me to understand it if you have solved? Thanks!

**Comment by dougbrion** [comment]
Thanks all have checked and reimplemented it as the following which I am pretty sure is correct. @iamwinter gamma(K)=1 <- this is not correct as K is the number of classes / size of the vector of ones.

Here is the equation from the paper:
![image](https://user-images.githubusercontent.com/17257556/128031213-e1a85a73-b264-4430-bdd3-dbdf5f345f99.png)

And the new implementation... let me know if you have any questions!

```python
def kl_divergence(alpha, num_classes, device=None):
    if not device:
        device = get_device()
    ones = torch.ones([1, num_classes], dtype=torch.float32, device=device)
    sum_alpha = torch.sum(alpha, dim=1, keepdim=True)
    first_term = (
        torch.lgamma(sum_alpha)
        - torch.lgamma(alpha).sum(dim=1, keepdim=True)
        + torch.lgamma(ones).sum(dim=1, keepdim=True)
        - torch.lgamma(ones.sum(dim=1, keepdim=True))
    )
    second_term = (
        (alpha - ones)
        .mul(torch.digamma(alpha) - torch.digamma(sum_alpha))
        .sum(dim=1, keepdim=True)
    )
    kl = first_term + second_term
    return kl
```

**Comment by Cebtenzzre** [comment]
@dougbrion As far as I can tell, the new kl_divergence implementation is functionally and numerically identical to the old one, with the only changes being refactoring and a few renamed variables. The differences in the NLL loss plot ([old](https://raw.githubusercontent.com/dougbrion/pytorch-classification-uncertainty/091a130b34716943119f83623854105482c6083c/results/rotate_uncertainty_log.jpg), [new](https://raw.githubusercontent.com/dougbrion/pytorch-classification-uncertainty/aeb739fdfc21c74638648c9b3af35d59bb329be1/results/rotate_uncertainty_log.jpg)) could be attributed to commit 5d5c0a62f05fd6d981b0a059712bcb2102daaded (new test digit) or the changes to data.py in commit c9d4845c97c6e5c0322bf4754274fc0f415fee0a (resize transform eliminated) - after all, the rotated digit is visibly different.

Is my analysis correct?

**Comment by dougbrion** [comment]
@Cebtenzzre yup you are correct.

Just re-wrote it to hopefully make it "clearer" and easier to understand for future users. All we are doing is breaking apart the log term of the equation in the paper.

The rest of your analysis is also correct. The differences are very minimal in the NLL loss plot - most likely due to just random training variance as seeds were not set for the trains. The resize makes no difference as MNIST data is already 28x28 pixels (don't know why I put it in in the first place).

**Comment by iamwinter** [comment]
I am sorry that I wrote an incorrect equation "gamma(K)=1" as I mistakenly thought K is a vector of ones. Thank you for pointing out my fault in patience and giving a clearer implement.

But I am still confused about the problem raised by @zhangchuanyi96 . Specifically, I think the code below is accord with the equation of the paper.
```
    first_term = (
        torch.lgamma(sum_alpha)
        - torch.lgamma(alpha).sum(dim=1, keepdim=True)
        - torch.lgamma(ones.sum(dim=1, keepdim=True))
    )
```
The part `+ torch.lgamma(ones).sum(dim=1, keepdim=True)` in your code which is always return a zero tensor, I can't find the corresponding part in the equation of the paper. Could you teach me why is it here?

Please excuse my ignorance. Thank you.
